txscript/sigcache.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2015 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package txscript
+
+import (
+	"bytes"
+	"crypto/rand"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// sigInfo represents an entry in the SigCache. Entries in the sigcache are a
+// 3-tuple: (sigCache, sig, pubKey).
+type sigInfo struct {
+	sigHash wire.ShaHash
+	sig     string
+	pubKey  string
+}
+
+// SigCache implements an ECDSA signature verification cache with a randomized
+// entry eviction policy. The benefits of SigCache are two fold. Firstly, usage
+// of SigCache mitigates a DoS attack wherein an attack causes a victim's client
+// to hang due to worst-case behavior triggered while processing attacker
+// crafted invalid transactions. A detailed description of the mitigated DoS
+// attack can be found here: https://bitslog.wordpress.com/2013/01/23/fixed-bitcoin-vulnerability-explanation-why-the-signature-cache-is-a-dos-protection/.
+// Secondly, usage of the SigCache introduces a signature verification
+// optimization which speeds up the validation of transactions within a block,
+// if they've already once been seen and verified within the mempool.
+type SigCache struct {
+	sync.RWMutex
+	validSigs  map[sigInfo]struct{}
+	maxEntries int
+}
+
+// NewSigCache creates and initializes a new instance of SigCache. Its sole
+// parameter 'maxEntries' represents the maximum number of entries allowed to
+// exist in the SigCache and any particular moment. If the addition of a new
+// 'sigInfo' entry puts the number of entries in the SigCache over
```

Comment:
Since `sigInfo` is an internal type, it probably doesn't make sense to make reference to it on the exported godoc.  Perhaps:

`Random entries are evicted to make room for new entries that would cause the number of entries in the cache to exceed the max.`